pkg/dns/outbound.go

```diff
@@ -28,12 +28,15 @@ func VIPOutbounds(
 
 		if dataplane.Spec.IsIngress() {
 			for _, service := range dataplane.Spec.Networking.Ingress.AvailableServices {
-				inService := service.Tags[mesh_proto.ServiceTag]
-				if _, found := serviceVIPMap[inService]; !found {
-					vip, err := ForwardLookup(vips, inService)
-					if err == nil {
-						serviceVIPMap[inService] = vip
-						services = append(services, inService)
+				if service.Mesh == dataplane.Meta.GetMesh() {
```

Comment:
It's not correct here to compare `service.Mesh` with `dataplane.Meta.GetMesh()` because this check happens when `dataplane` variable points to the Ingress, but in reality is not mesh-scoped. Instead, you have to compare `service.Mesh` against the mesh of the dataplane with name `name`. 

So I'd change the signature of VIPOutbounds to receive `dpKey` instead of just `name`:
```go
func VIPOutbounds(
	dpKey model.ResourceKey,
	dataplanes []*core_mesh.DataplaneResource,
	vips vips.List,
	externalServices []*core_mesh.ExternalServiceResource,
) []*mesh_proto.Dataplane_Networking_Outbound {
...
    if dataplane.Spec.IsIngress() {
        for _, service := range dataplane.Spec.Networking.Ingress.AvailableServices {
            if service.Mesh == dpKey.Mesh {
            ...
}
```
What do you think about it?

Reply:
All dataplanes in `dataplanes` are from the same Mesh so I believe it doesn't make much of a difference.
Anyhow, If it helps with readability I'll update it.

Reply:
List `dataplanes` consists of the dataplanes from the same mesh + Ingresses which effectively could belong to any mesh. This is how we generate `dataplanes` list on Kubernetes, [code](https://github.com/kumahq/kuma/blob/master/pkg/plugins/runtime/k8s/controllers/pod_controller.go#L185):

```go
otherDataplanes := make([]*mesh_k8s.Dataplane, 0)
for i := range allDataplanes.Items {
	dataplane := allDataplanes.Items[i]
        ...
	if dataplane.Mesh == mesh || dp.Spec.IsIngress() {
		otherDataplanes = append(otherDataplanes, &dataplane)
	}
}
```
and on Universal we do it wrong (this is another bug), [code](https://github.com/kumahq/kuma/blob/master/pkg/plugins/runtime/universal/outbound/outbound.go#L68):
```go
dpList := &mesh.DataplaneResourceList{}
if err := v.rorm.List(ctx, dpList, store.ListByMesh(m.Meta.GetName())); err != nil {
	return err
}
```
So on Universal we also should include Ingresses into `dataplanes` list and they could have different mesh. That's why it's important to check `service.Mesh` against `dpKey.Mesh` rather than `dataplane.Meta.GetMesh()`.

Reply:
Ha right I got tricked by this other bug. Thanks a lot for the explanation it makes perfect sense indeed (You need to check all ingresses to get all available services).
I'll fix this here as this patch won't work without this.

Btw this makes this scale in total number of dp rather than number of dps in a mesh?
Unfortunately while k8s forces ingresses to be in the "default" mesh I don't believe this is the case for universal mode.
So we need to check all meshes to get all ingresses doesn't it?
I guess forcing ingresses in "default" for universal would be a breaking change and probably not something we'd want to do no?

Finally, ingresses will be a scaling bottleneck anyway as it more or less needs to know about every service in any mesh.
Are there other places where you'd envision complexity in the order of total number of services rather than number of services inside a mesh?

Reply:
Hey @lobkovilya I've updated the PR to fix this second bug. There were multiple ways to go about it, I picked what looked to be the most readable to me.